docs/side_quests/nf-test.md

```diff
@@ -0,0 +1,832 @@
+
+Maybe you add a simple true or false parameter such as `--skip_process`, now you must run the pipeline twice, once with each parameter to make sure it works as expected. But wait, how do we check if the `--skip_process` actually skips the process? We have to dig into the outputs or check the log files! This is a pain and prone to error.
+
+As you develop your pipeline, it will quickly become so complex that manually testing every iteration is slow and error prone. Furthermore, if you do find an error it will be very difficult to pin down exactly where in your pipeline the error is coming from. This is where unit testing comes in.
+
+Testing allows you to systematically check that every part of your pipeline is working as expected. The benefits to a developer are huge:
+
+- **Confidence**: Because the tests cover the entire pipeline, you can be be confident changing something doesn't affect anything else
+- **Trust**: When multiple developers work on the pipeline, they know the other developers haven't broken the pipeline and every component.
+- **Transparency**: The tests show where a pipeline is failing and make it easier to track down the problem. They also function as a form of documentation, showing how to run a process or workflow.
+- **Speed**: Because the tests are automated, they can be run very quickly and repeatedly. You can iterate quickly with less fear of introducing new bugs.
+
+There are lots of different types of tests we can write:
+
+1. **Module-level tests**: For individual processes
+2. **Workflow-level tests**: For a single workflow
+3. **Pipeline-level tests**: For the pipeline as a whole
+4. **Integration tests**: For the interaction of the pipeline with other systems
+5. **Performance tests**: For the speed and efficiency of the pipeline
+6. **Stress tests**: To identify the limits of the pipeline
+
+[**nf-test**](https://www.nf-test.com/) is a tool that allows you to write module, workflow and pipeline level test. In short, it allows you to systematically check every individual part of the pipeline is working as expected, _in isolation_.
+
+In this part of the training, we're going to show you how to use nf-test to write module-level tests for the three processes in our pipeline.
+
+---
+
+## 0. Warmup
+
+Let's move into the project directory.
+
+```bash
+cd side-quests/nf-test
+```
+
+The `nf-test` directory has the file content like:
+
+```console title="Directory contents"
+nf-test
+├── greetings.csv
+├── main.nf
+└── nextflow.config
+```
+
+For a detailed description of the files, see the [warmup from Hello Nextflow](../hello_nextflow/00_orientation.md).
+
+### 0.1. Run the workflow
+
+Let's run the workflow to make sure it's working as expected.
+
+```bash
+nextflow run main.nf
+```
+
+CONGRATULATIONS! You just ran a unit test!
+
+"Wait, what? I just ran the workflow and it worked! How is that a unit test?"
```

Comment:
This is confusing for a couple of reasons. 

1. 'Unit test' is the only type of test not explicitly listed above, and non-programmers will typically have never heard of unit tests.
2. Running an entire workflow is not like a unit test at all; arguably the closest thing to a unit test here is a process-level test.

FWIW when I taught the original nf-test material I made the approximation that process level tests are like unit tests, and that workflow level tests are like integration tests (because they test the integration of all the processes together). I'd like to use that formally in these materials.

Reply:
Semantics is everything when we're deciding what to call things 😆 

I'm going off of what the GATK devs called integration tests, but recognize different groups use different terminologies.

So ok, let's just call this a workflow-level test or something to that effect.

Reply:
I'm with @vdauwera here actually- testing processes feels like unit testing, workflows are integration tests. 

But maybe avoiding the language altogether and just being specific on what we mean is the best way forward in any case.

Reply:
Just 'test' is fine here, changed accordingly